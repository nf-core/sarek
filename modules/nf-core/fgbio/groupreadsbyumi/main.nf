process FGBIO_GROUPREADSBYUMI {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/fgbio:2.2.1--hdfd78af_0' :
        'biocontainers/fgbio:2.2.1--hdfd78af_0' }"

    input:
    tuple val(meta), path(bam)
    val(strategy)

    output:
    tuple val(meta), path("*.bam")         , emit: bam
    tuple val(meta), path("*histogram.txt"), emit: histogram
    path "versions.yml"                    , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}_umi-grouped"
    def mem_gb = 8
    if (!task.memory) {
        log.info '[fgbio FilterConsensusReads] Available memory not known - defaulting to 8GB. Specify process memory requirements to change this.'
    } else if (mem_gb > task.memory.giga) {
        if (task.memory.giga < 2) {
            mem_gb = 1
        } else {
            mem_gb = task.memory.giga - 1
        }
    }

    if ("$bam" == "${prefix}.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

    """
    fgbio \\
        -Xmx${mem_gb}g \\
        --tmp-dir=. \\
        GroupReadsByUmi \\
        -s $strategy \\
        $args \\
        -i $bam \\
        -o ${prefix}.bam \\
        -f ${prefix}_histogram.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}_umi-grouped"
    if ("$bam" == "${prefix}.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    """
    touch ${prefix}.bam
    touch ${prefix}_histogram.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
    END_VERSIONS
    """
}
