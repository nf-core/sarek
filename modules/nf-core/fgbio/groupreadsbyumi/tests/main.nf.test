nextflow_process {

    name "Test Process FGBIO_GROUPREADSBYUMI"
    script "../main.nf"
    process "FGBIO_GROUPREADSBYUMI"

    tag "modules"
    tag "modules_nfcore"
    tag "fgbio"
    tag "fgbio/groupreadsbyumi"

    test("sarscov2 - bam") {

        when {
            process {
                """
                input[0] = [
                            [ id:'test', single_end:false ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/umi/test.paired_end.unsorted_tagged.bam', checkIfExists: true)
                        ]
                input[1] = "Adjacency"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - bam - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                            [ id:'test', single_end:false ], // meta map
                            file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/umi/test.paired_end.unsorted_tagged.bam', checkIfExists: true)
                        ]
                input[1] = "Adjacency"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
