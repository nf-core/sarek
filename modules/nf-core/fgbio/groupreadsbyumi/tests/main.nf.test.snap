{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped_histogram.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,e293eed3614f921114b3bd5b0e1ada10"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "histogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped_histogram.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e293eed3614f921114b3bd5b0e1ada10"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-21T10:48:29.108067677"
    },
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped.bam:md5,35bfc992c30d8e3e50816159fa58cb11"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped_histogram.txt:md5,9a0c622b65209afbce0840e2affff983"
                    ]
                ],
                "2": [
                    "versions.yml:md5,e293eed3614f921114b3bd5b0e1ada10"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped.bam:md5,35bfc992c30d8e3e50816159fa58cb11"
                    ]
                ],
                "histogram": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_umi-grouped_histogram.txt:md5,9a0c622b65209afbce0840e2affff983"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e293eed3614f921114b3bd5b0e1ada10"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-05-21T10:48:14.269677258"
    }
}